python: don't use keep for hooks, so they will be spliced

Artturin opened the pull request:
See https://github.com/NixOS/nixpkgs/issues/211340 for prerequisite 


The wrapPython weirdness is to fix
```
error: output '/nix/store/li07bad6s9brxx7cf8j408631nca39ka-python3.10-chardet-5.1.0-aarch64-unknown-linux-gnu' is not allowed to refer to the following paths:
         /nix/store/fdqpyj613dr0v1l1lrzqhzay7sk4xg87-python3-3.10.10
```

ericson said in https://github.com/NixOS/nixpkgs/pull/104201#discussion_r527041018 that the reason they were in keep was to not splice them but before the commit which broke hook splicing (33d12e5) everything appeared to be working correctly when cross-compiling


###### Description of changes

experimenting with hook splicing fix @amjoseph-nixpkgs 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin commented on pkgs/development/interpreters/python/default.nix:
```diff
@@ -61,7 +61,7 @@
           };
           hooks = import ./hooks/default.nix;
-          keep = lib.extends hooks pythonPackagesFun;
+          keep = self: { inherit (self) wrapPython; };
```
would be idea to splice wrapPython too so i could move forward with https://github.com/NixOS/nixpkgs/pull/204387

Artturin replied:
done

Artturin committed:
python: don't use keep for hooks, so they will be spliced

as far as i can see there should be no reason to not splice the hooks

wrapPython for some reason has to be not spliced because 'substitutions.pythonHost = python;' causes
```
error: output '/nix/store/li07bad6s9brxx7cf8j408631nca39ka-python3.10-chardet-5.1.0-aarch64-unknown-linux-gnu' is not allowed to refer to the following paths:
         /nix/store/fdqpyj613dr0v1l1lrzqhzay7sk4xg87-python3-3.10.10
```

ericson said in #104201 (comment) that the reason they were in keep was to not splice them but before the commit which broke hook splicing (33d12e5) everything appeared to be working correctly when cross-compiling

we also somehow get a correctly spliced python in nativeBuildInputs for free

Artturin committed:
python3Packages.wrapPython: Splice

Using __spliced is the easiest (maybe only) way to get the correct python here.

wrapPython goes in nativeBuildInputs.

By adding `substitutions.python1 = python;` to wrap-python.nix we can
see that hostHost contains a python for aarch64 but not hostTarget for
some reason.

```
nix-repl> p = lib.elemAt pkgsCross.aarch64-multiplatform.__splicedPackages.python3Packages.chardet.nativeBuildInputs 1
nix-repl> p.python1.__spliced
{ buildBuild = «derivation /nix/store/21ldw2dp26xvv9iyxn9x77a8yh4waqz5-python3-3.10.12.drv»; buildHost = «derivation /nix/store/21ldw2dp26xvv9iyxn9x77a8yh4waqz5-python3-3.10.12.drv»; buildTarget = «derivation /nix/store/21ldw2dp26xvv9iyxn9x77a8yh4waqz5-python3-3.10.12.drv»; hostHost = «derivation /nix/store/39dkb51rciw6zwg0c2c44gpmpjapddxc-python3-aarch64-unknown-linux-gnu-3.10.12.drv»; hostTarget = «derivation /nix/store/21ldw2dp26xvv9iyxn9x77a8yh4waqz5-python3-3.10.12.drv»; }

```

Artturin commented:
@Ericson2314 maybe made them not splice to work around https://github.com/NixOS/nixpkgs/issues/211340
![image](https://github.com/NixOS/nixpkgs/assets/56650223/d46a80a1-be6d-4007-ba22-235750965b91)

https://github.com/NixOS/nixpkgs/blob/7482a0e6ef6c53539467df143d825a9545eacf67/pkgs/development/tools/misc/itstool/default.nix#L20


```
$ nix build ".#pkgsCross.aarch64-multiplatform.itstool"
error:
       … while calling the 'derivationStrict' builtin

         at <nix/derivation-internal.nix>:9:12:

            8|
            9|   strict = derivationStrict drvAttrs;
             |            ^
           10|

       … while evaluating derivation 'itstool-aarch64-unknown-linux-gnu-2.0.6'
         whose name attribute is located at /home/artturin/nixgits/my-nixpkgs/pkgs/stdenv/generic/make-derivation.nix:300:7

       … while evaluating attribute 'nativeBuildInputs' of derivation 'itstool-aarch64-unknown-linux-gnu-2.0.6'

         at /home/artturin/nixgits/my-nixpkgs/pkgs/stdenv/generic/make-derivation.nix:344:7:

          343|       depsBuildBuild              = lib.elemAt (lib.elemAt dependencies 0) 0;
          344|       nativeBuildInputs           = lib.elemAt (lib.elemAt dependencies 0) 1;
             |       ^
          345|       depsBuildTarget             = lib.elemAt (lib.elemAt dependencies 0) 2;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: makeWrapper/makeShellWrapper must be in nativeBuildInputs
```

amjoseph-nixpkgs commented:
Let me be totally honest here, I'm trying to remove setup hooks from nixpkgs -- starting from the `cc-wrapper` and `binutils-wrapper` which is where this whole mess begins.

The main reason why setup hooks exist is that there are pieces of data that get stuffed into `$out/nix-support` instead of in Nix attrsets.  Once that happens, you can't branch/select/switch on those values at eval time (without IFD of course).  When these values are used to compute other values, you get an "avalanche" of logic that gets forced out of the Nix eval language and into hideous gross piles of `bash`.

It turns out that 99% of this stuff is actually knowable at Eval time.  But you have to stop the avalanche before it starts.

That is my mission.

With all that said, @Artturin I appreciate your heroic efforts to tame the bash-monster, and we need these efforts.  I just wanted to explain why it's really hard for me to motivate myself to get a deep enough understanding of problems with this stuff on one hand when I'm putting a lot of effort into being able to get rid of it on the other.

It's also starting to look like I'm probably the only even-semi-qualified reviewer for a lot of this stuff.  I will give it a try.  Please keep pestering me.

I'm going to cherry-pick this into my local tree and rebuild my 100%-cross platforms (aarch64 and mips) overnight to see what breaks.

FRidh commented:
> It turns out that 99% of this stuff is actually knowable at Eval time. But you have to stop the avalanche before it starts.

This is interesting. I've been wanting this as well, however, a big worry is that eval time increases. 

E.g., with Python we still use propagated-build-inputs. We can get rid of that in two ways, either writing a custom file linking runtime deps, or keeping track of deps in eval only and writing references out when building an env.

amjoseph-nixpkgs commented:
Addendum: the kicker starts with this commit, which isn't part of the PR I linked (but goes immediately on top of it).  This commit does not eval, but it's sort of the first milestone I've been working towards:

- https://github.com/amjoseph-nixpkgs/nixpkgs/commit/3d45166c6e46d8a3732532745b07da4d97029bc4

> keeping track of deps in eval only and writing references out when building an env

You definitely want to do this.

> I've been wanting this as well, however, a big worry is that eval time increases.

Why would it though?  The evaluator still has to instantiate all ancestral dependencies either way.  So the number of derivations instantiated is the same either way; the only difference is whether the list is passed as lines in `nix-support/propagated-build-inputs` or as a Nix list.

It might make sense to "keep track of deps" using an attrset rather than a list though, since attrset union is `O(n)` (cppnix implements this by merging pre-sorted lists) but list-concatenation is (in the pathological case) `O(n^2)`.

amjoseph-nixpkgs commented:
This breaks eval for `claws-mail` but everything else was fine.  Builds underway; I'll check back in the morning.

Built on:
- [ ] `aarch64-linux` laptop packageset (cross from `x86_64-linux`)
- [ ] `mips64el-linux` router packageset (cross from `x86_64-linux`)
  - [ ] `abi=64` (stuff that uses `rust`, `golang`, and `boost-context`)
  - [ ] `abi=n32` (everything else)

Artturin commented:
useful
`nix-diff $(nix eval --raw "nixpkgs/$(git merge-base staging HEAD)#pkgsCross.aarch64-multiplatform.python3Packages.frozenlist.drvPath") $(nix eval --raw ".#pkgsCross.aarch64-multiplatform.python3Packages.frozenlist.drvPath")`

Artturin commented:
`python3Packages.cryptography` cross will be fixed by https://github.com/NixOS/nixpkgs/pull/212795, so not a regression

the other 3 build on https://github.com/NixOS/nixpkgs/pull/245475 so i'm going to merge that pr

amjoseph-nixpkgs commented:
<details><summary>PyQt5 failure</summary>
<pre>
error: builder for '/nix/store/w8a7pipahqqj1mg30kdvh33ilr88m7bk-python3.10-PyQt5-5.15.9-aarch64-unknown-linux-gnu.drv' failed with exit code 1;
       last 10 log lines:
       >   full command: /nix/store/ci5b14j87px5h2j83244qrdxzy9c89lc-python3-3.10.12/bin/python3.10 /nix/store/04r3170l9d81id0gd4nk1cr14jjbfa3p-python3.10-pip-23.0.1/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py prepare_metadata_for_build_wheel /build/tmpxoo9ltc7
       >   cwd: /build/PyQt5-5.15.9
       >   Preparing metadata (pyproject.toml) ... error
       > error: metadata-generation-failed
       >
       > × Encountered error while generating package metadata.
       > ╰─> See above for output.
       >
       > note: This is an issue with the package mentioned above, not pip.
       > hint: See above for details.
       For full logs, run 'nix log /nix/store/w8a7pipahqqj1mg30kdvh33ilr88m7bk-python3.10-PyQt5-5.15.9-aarch64-unknown-linux-gnu.drv'.

</pre>
</details>

amjoseph-nixpkgs commented:
> `python3Packages.cryptography` cross will be fixed by #212795, so not a regression

I already have the equivalent of that PR in my tree.  `python3Packages.cryptography` cross-compiled before cherry-picking this PR, and failed after.

Granted I'm not using exactly #212795's commits, but that PR has been such a headache to review; the author doesn't seem to believe in using the "commit suggestion" button, or even putting each response in a separate commit.  They just squash+forcepush making the reviewers re-review the whole thing.  It's kind of annoying.

amjoseph-nixpkgs commented:
Do you think we could merge my qutebrowser cross-compilation PRs ahead of this one?  At least that way we could get ofborg to verify that qutebrowser builds before merging this (or some future version of this) as well as after (i.e. no regressions).

They've been pending since ~April-ish.

Specifically,

- https://github.com/NixOS/nixpkgs/pull/227900
- https://github.com/NixOS/nixpkgs/pull/227902
- https://github.com/NixOS/nixpkgs/pull/227901
- https://github.com/NixOS/nixpkgs/pull/220315
- https://github.com/NixOS/nixpkgs/pull/220319
- https://github.com/NixOS/nixpkgs/pull/220314
- https://github.com/NixOS/nixpkgs/pull/220316

amjoseph-nixpkgs commented:
> I already have the equivalent of that PR in my tree. `python3Packages.cryptography` cross-compiled before cherry-picking this PR, and failed after.

Sorry it was not this PR, but rather #245475 which [breaks #212795](https://github.com/NixOS/nixpkgs/pull/245475#issuecomment-1652686302).  I cherry-picked both before running the rebuild.

Artturin committed:
python3Packages.wrapPython: use targetTarget

For whatever reason after 1 year `targetTarget` now contains cross
python while in the previous commit (commit msg from it)

By adding `substitutions.python1 = python;` to wrap-python.nix we can
see that hostHost contains a python for aarch64 but not hostTarget for
some reason.

```
nix-repl> p = lib.elemAt pkgsCross.aarch64-multiplatform.__splicedPackages.python3Packages.chardet.nativeBuildInputs 1
nix-repl> p.python1.__spliced
{ buildBuild = «derivation /nix/store/21ldw2dp26xvv9iyxn9x77a8yh4waqz5-python3-3.10.12.drv»; buildHost = «derivation /nix/store/21ldw2dp26xvv9iyxn9x77a8yh4waqz5-python3-3.10.12.drv»; buildTarget = «derivation /nix/store/21ldw2dp26xvv9iyxn9x77a8yh4waqz5-python3-3.10.12.drv»; hostHost = «derivation /nix/store/39dkb51rciw6zwg0c2c44gpmpjapddxc-python3-aarch64-unknown-linux-gnu-3.10.12.drv»; hostTarget = «derivation /nix/store/21ldw2dp26xvv9iyxn9x77a8yh4waqz5-python3-3.10.12.drv»; }

```

NOW

nix-repl> pkgsCross.aarch64-multiplatform.buildPackages.python3Packages.wrapPython.python1.__spliced
{
  buildBuild = «derivation /nix/store/5k1b6g2qgy2a7kiis6snyidzryz6bljl-python3-3.12.5.drv»;
  buildHost = «derivation /nix/store/5k1b6g2qgy2a7kiis6snyidzryz6bljl-python3-3.12.5.drv»;
  buildTarget = «derivation /nix/store/5k1b6g2qgy2a7kiis6snyidzryz6bljl-python3-3.12.5.drv»;
  hostHost = «derivation /nix/store/5k1b6g2qgy2a7kiis6snyidzryz6bljl-python3-3.12.5.drv»;
  hostTarget = «derivation /nix/store/5k1b6g2qgy2a7kiis6snyidzryz6bljl-python3-3.12.5.drv»;
  targetTarget = «derivation /nix/store/iqjs63kixgafzs41jkc8m8gizvjwl1lh-python3-aarch64-unknown-linux-gnu-3.12.5.drv»;
}

note that previously there wasn't even a `targetTarget`

Artturin committed:
python3Packages: hook deps to next offset

fixes

```
python3.12-packaging> /nix/store/wk4c4cjrrssz3vrq3pxh04wmzd3d93y6-python-runtime-deps-check-hook.sh/nix-support/setup-hook: line 11: /nix/store/vc4xlrlswjpwfjcd1f018ykpdd3rla8i-python3-aarch64-unknown-linux-gnu-3.12.5/bin/python3.12: cannot execute binary file: Exec format error
```

`.override` discards splicing

Artturin commented:
Rebased but there's some serious issues with python cross now as not even this can make `python` and the hooks spliced
so the message in the first commit `we also somehow get a correctly spliced python in nativeBuildInputs for free` is now wrong 

```
nix-repl> pkgsCross.aarch64-multiplatform.flashfocus.nativeBuildInputs
[
  «derivation /nix/store/5k1b6g2qgy2a7kiis6snyidzryz6bljl-python3-3.12.5.drv»
  «derivation /nix/store/fx72jxhykc327pd3lym03x9ajprm471d-wrap-python-hook.drv»
  «derivation /nix/store/alb6sn9w7gvrh50cnh3q6wnrnvwhi841-ensure-newer-sources-hook.drv»
  «derivation /nix/store/yzwjwd5arv6vpzfdxrgsp750rbxwx2iq-python-remove-tests-dir-hook.drv»
  «derivation /nix/store/0ngx73bzdilmp1zn151kb1xgia5jyn7h-python-relax-deps-hook.drv»
  «derivation /nix/store/9sl7kpk8playqdi8i2vmhp90mkw04iis-python-remove-bin-bytecode-hook.drv»
  «derivation /nix/store/kv0zz11pfcr21yp8xhv4pkhkv3qa7ng9-pypa-build-hook.sh.drv»
  «derivation /nix/store/6dpb1rwmvxfx2g72qszsgfma6frcnkhh-python-runtime-deps-check-hook.sh.drv»
  «derivation /nix/store/6lp9xisrwzkmvgac6vnjnil29g2ii25b-pypa-install-hook.drv»
  «derivation /nix/store/fwyagzwkj49b8y3hj468qia68rdl5xqj-python-namespaces-hook.sh.drv»
  «derivation /nix/store/h05097ifb6kjab4fiin7gfzkxwq3xl10-python-output-dist-hook.drv»
  «derivation /nix/store/1ay4kcjvb78xr2z83r3c5rqz8dik5z58-python3.12-setuptools-72.1.0.drv»
]

nix-repl> p = lib.elemAt pkgsCross.aarch64-multiplatform.flashfocus.nativeBuildInputs 0

nix-repl> p ? __spliced
false

nix-repl> p = lib.elemAt pkgsCross.aarch64-multiplatform.flashfocus.nativeBuildInputs 1

nix-repl> p ? __spliced
false
```

Artturin committed:
python3Packages.wrapPython: hostHost is still correct

`wrap-python.nix`

```
passthru = {
  python1 = python;
};
```

```
nix-repl> p = lib.elemAt pkgsCross.aarch64-multiplatform.flashfocus.nativeBuildInputs 1

nix-repl> p.python1.__spliced
{
  buildBuild = «derivation /nix/store/5k1b6g2qgy2a7kiis6snyidzryz6bljl-python3-3.12.5.drv»;
  buildHost = «derivation /nix/store/5k1b6g2qgy2a7kiis6snyidzryz6bljl-python3-3.12.5.drv»;
  buildTarget = «derivation /nix/store/5k1b6g2qgy2a7kiis6snyidzryz6bljl-python3-3.12.5.drv»;
  hostHost = «derivation /nix/store/iqjs63kixgafzs41jkc8m8gizvjwl1lh-python3-aarch64-unknown-linux-gnu-3.12.5.drv»;
  hostTarget = «derivation /nix/store/5k1b6g2qgy2a7kiis6snyidzryz6bljl-python3-3.12.5.drv»;
}

```

Artturin committed:
try

doesn't work

`pkgsCross.aarch64-multiplatform.flashfocus`

```
> /nix/store/lxmqqrafnjnjkdplvsk2yy6fx5f5g00z-python3.12-build-1.2.1/bin/.pyproject-build-wrapped: /nix/store/lxmqqrafnjnjkdplvsk2yy6fx5f5g00z-python3.12-build-1.2.1/bin/pyproject-build: line 3: syntax error near unexpected token `lambda'
       > /nix/store/lxmqqrafnjnjkdplvsk2yy6fx5f5g00z-python3.12-build-1.2.1/bin/.pyproject-build-wrapped: /nix/store/lxmqqrafnjnjkdplvsk2yy6fx5f5g00z-python3.12-build-1.2.1/bin/pyproject-build: line 3: `import sys;import site;import functools;sys.argv[0] = '/nix/store/lxmqqrafnjnjkdplvsk2yy6fx5f5g00z-python3.12-build-1.2.1/bin/pyproject-build';functools.reduce(lambda k, p: site.addsitedir(p, k), ['/nix/store/lxmqqrafnjnjkdplvsk2yy6fx5f5g00z-python3.12-build-1.2.1/lib/python3.12/site-packages','/nix/store/gmldyqksmb8m3q55r89plzl69i2hpn1z-python3.12-packaging-24.1/lib/python3.12/site-packages','/nix/store/h3i0acpmr8mrjx07519xxmidv8mpax4y-python3-3.12.5/lib/python3.12/site-packages','/nix/store/85r0faqh44wfff7250qxad4iif2kng05-python3.12-pyproject-hooks-1.0.0/lib/python3.12/site-packages'], site._init_pathinfo());'
       For full logs, run 'nix log /nix/store/dcygh555yx6sqzqrlnnd0xdq1bflycsk-python3.12-setuptools-72.1.0.drv'.
error: builder for '/nix/store/4mzla40czanjwsg3swqgviyzc342j1mq-python3.12-typing-extensions-4.12.2.drv' failed with exit code 2;
```

Artturin commented:
Yeah idk what's going on, doesn't work
